docs/custom-filters.adoc

```diff
@@ -407,75 +419,57 @@ public class DisallowAlterConfigs implements AlterConfigsRequestFilter {
+
+The asynchronous request/response will be intercepted by Filters upstream of this Filter.  Filters downstream of this
+Filter (and the Client) do not see the asynchronous response.
+
+Let's take a look at an example. We'll send an asynchronous request towards the Cluster for topic metadata while
+handling a FetchRequest and use the response to mutate the FetchRequest before passing it to the next filter in the chain.
 
 [source,java]
 ----
-public class FetchFilter implements FetchRequestFilter{
+public class FetchFilter implements FetchRequestFilter {
     public static final short METADATA_VERSION_SUPPORTING_TOPIC_IDS = (short) 12;
 
     @Override
-    public CompletionStage<RequestFilterResult> onRequest(ApiKeys apiKey,
-                                                          RequestHeaderData header,
-                                                          ApiMessage body,
-                                                          KrpcFilterContext filterContext) {
+    public CompletionStage<RequestFilterResult> onFetchRequest(ApiKeys apiKey,
+                                                               RequestHeaderData header,
+                                                               FetchRequestData request,
+                                                               KrpcFilterContext context) {
         MetadataRequestData metadataRequest = new MetadataRequestData(); //<1>
         var topic = new MetadataRequestData.MetadataRequestTopic();
         topic.setTopicId(Uuid.randomUuid());
         metadataRequest.topics().add(topic);
         short version = METADATA_VERSION_SUPPORTING_TOPIC_IDS;
-        CompletionStage<MetadataResponseData> stage = context.sendRequest(version, request);//<2>
-        stage.thenAccept(response -> doSomethingWithMetadata(response, version));//<3>
-        return context.forwardRequest(header, request);//<4>
+        CompletionStage<MetadataResponseData> stage = context.sendRequest(version, request); //<2>
+        return stage.thenApply(metadataResponse -> mutateFetchRequest(metadataResponse, request)) // <3>
+                    .thenCompose(mutatedFetchRequest -> context.forwardRequest(header, mutatedFetchRequest)); // <4>
     }
 }
 ----
-<1> We construct a new request object
+<1> We construct a new request object.  This will be the asynchronous request.
 <2> We send the request towards the Cluster, specifying the api version to use and obtaining a CompletionStage for the
 response.
-<3> When the CompletionStage is completed, we do something with the metadata (populate a cache for example). Note that this
-stage is completed on the same thread as the rest of the Filter work, so we can safely mutate Filter members without
-synchronising.
-<4> We also forward the original request. This will flow to the Cluster and back through Kroxylicious to the Client.
+<3> We use CompletionStage chaining to mutate the filtered fetch request using the asynchronous response.
```

Comment:
```suggestion
<3> We use CompletionStage chaining to mutate the filtered fetch request using the response from the request sent at <1>.
```

Reply:
https://github.com/asciidoctor/asciidoctor/issues/1077

There isn't a supported way to reference the callouts (so there is no fancy hyperlinking), but what you have done seems to render in a reasonable way.